ppx/ppx_cstruct.ml

```diff
@@ -157,21 +181,55 @@ let op_evar s op = Ast.evar (op_name s op)
 let get_expr loc s f =
   let m = mode_mod loc s.endian in
   let num x = Ast.int x in
+  let off = f.off / 8 in
   match f.ty with
   |Buffer (_, _) ->
-    let len = width_of_field f in
-    [%expr
-      fun src -> Cstruct.sub src [%e num f.off] [%e num len]
-    ]
+    let len = (width_of_field f) / 8 in
+    [%expr fun src -> Cstruct.sub src [%e num off] [%e num len]]
+  |Bitfield (prim, len) ->
+    (* It depends on the actual bits count, so if the bits count
+     * lesser than primary type size - we still take primary type,
+     * but if it is bigger - we take the bigger unit then *)
+    if len > 0 && len <= 8 then
+    [%expr fun v ->
+          [%e match prim with
+              |Char |UInt8 ->
+                [%expr [%e m "get_bits_uint8"] v [%e num f.off] [%e num len]]
```

Comment:
These bits can be factored out by computing the getter function for each prim type.